### PR TITLE
feat: implement transitive dependencies

### PR DESCRIPTION
**Pull request by pablo-snyk**
In order to extend our vulnerability scanning feature to support child dependencies for [npm packages](https://docs.npmjs.com/cli/v6/configuring-npm/package-json) as described in #5  , this pull request updates the package service to return all nested dependencies on the internal `/package` endpoint for consumption by vulnerability service instead of just returning the versions for the first level.

It supports a `GET` request to the `/package/:name/:version` endpoint and will return a JSON structure representing the full tree of dependencies. We will always return the latest matching version of a package supported to mimic the behavior of a fresh `npm install`.

**Testing**

It can be tested locally by making a request for a package with sub-dependencies e.g. `react@16.13.0`

    curl -s http://localhost:3000/package/react/16.13.0 | jq .

**Related Ticket**

* #5

### CODE REVIEW

File: packages/models.py [hunk 1]
```diff
@@ -1,6 +1,12 @@
 class VersionedPackage:
-    def __init__(self, name: str, version: str, description: str, dependencies: list[tuple]):
+    def __init__(
+        self,
+        name: str,
+        version: str,
+        description: str,
+        dependencies: list["VersionedPackage"] | None = None
```
Comment (alexbrebu) [inline packages/models.py:7]: Using "VersionedPackage" instead of "tuple" allows for more complex and recursive structures.

Comment (alexbrebu) [reply packages/models.py:7]: Allowing None and setting an [] when None is provided ensures that `self.dependencies` is always a list, simplifying usage and avoiding potential errors from passing None.

Comment (rachelcdavies) [reply packages/models.py:7]: @alexbrebu this repository is for tech test code review exercises which we give ti candidates who want to work for Snyk. So the code deliberately contains areas that we hope a candidate who is applying to work may suggest to improve. 
If we post perfect clean code it will make it harder for us to use the candidates in tech tests.
If you are a candidate then we invite you to a session to show us your findings, posting on the PR makes your comments available to other candidates.
I hope you don't mind us deleting your comments for the above reasons cc @pablo-snyk

File: packages/modules/npm.py [hunk 2]
```diff
@@ -1,21 +1,44 @@
...
-def get_versioned_package(name: str, version: str) -> VersionedPackage:
-    url = f"{NPM_REGISTRY_URL}/{name}/{version}"
+def get_package(name: str, range: str) -> VersionedPackage:
+    url = f"{NPM_REGISTRY_URL}/{name}"
 
-    response = requests.get(url)
-    npm_package = response.json()
+    npm_package = requests.get(url).json()
+    versions = list(npm_package["versions"].keys())
+    version = semver.min_satisfying(versions, range)
```
Comment (avivna) [inline packages/modules/npm.py:14]: Using semver.min_satisfying may return a deprecated package version (e.g. express package). i.e. you will need first to filter out all deprecated versions and apply min_satisfying search on the filtered list.

File: packages/modules/npm.py [hunk 3]
```diff
@@ -1,21 +1,44 @@
...
+
+    package.dependencies = [
+        get_package(name=dep_name, range=dep_range) for dep_name, dep_range in dependencies.items()
+    ]
 
     return package
+
+
+def request_package(name: str, range: str) -> tuple[VersionedPackage, dict]:
+    url = f"{NPM_REGISTRY_URL}/{name}"
```
Comment (avivna) [inline packages/modules/npm.py:32]: There is a significant code duplication here. common logic in get_package and request_package can be extracted to a util/helper function.

File: packages/modules/npm.py [hunk 4]
```diff
@@ -1,21 +1,44 @@
...
 
 
-def get_versioned_package(name: str, version: str) -> VersionedPackage:
-    url = f"{NPM_REGISTRY_URL}/{name}/{version}"
+def get_package(name: str, range: str) -> VersionedPackage:
+    url = f"{NPM_REGISTRY_URL}/{name}"
 
-    response = requests.get(url)
-    npm_package = response.json()
+    npm_package = requests.get(url).json()
```
Comment (avivna) [inline packages/modules/npm.py:12]: blocking: Need to add error handling (log + throw exception) in case package was not found, or server returns an error code

File: packages/modules/npm.py [hunk 5]
```diff
@@ -1,21 +1,44 @@
...
-        version=npm_package["version"],
-        dependencies=npm_package["dependencies"],
+        name=version_record["name"],
+        version=version_record["version"],
+        description=version_record["description"],
     )
+    dependencies = version_record.get("dependencies", {})
+
+    package.dependencies = [
+        get_package(name=dep_name, range=dep_range) for dep_name, dep_range in dependencies.items()
```
Comment (avivna) [inline packages/modules/npm.py:25]: Blocking: In case of dependency loop, infinite recursive calls may occur, that will eventually cause stack overflow.  For each recursive call, it is advisable to pass a set of all parent packages, and for each package - check if is already found in the set of parents - which indicates on a loop. In such case, exception should be thrown.

Non-blocking: package may include multiple dependencies, which in turn can include multiple dependencies. Has a result, packages fetching time may be long. Consider use asyncio coroutines the make concurrent calls to "get_package". Note this should be planned carefully, as you will need to verify all subtasks for a specific task/package  are completed before carrying on with the task, and you'd also need to handle concurrent cache updates.

File: packages/modules/npm.py [hunk 6]
```diff
@@ -1,21 +1,44 @@
...
+    version_record = npm_package["versions"][version]
 
     package = VersionedPackage(
-        name=npm_package["name"],
-        description=npm_package["description"],
-        version=npm_package["version"],
-        dependencies=npm_package["dependencies"],
+        name=version_record["name"],
+        version=version_record["version"],
+        description=version_record["description"],
```
Comment (avivna) [inline packages/modules/npm.py:20]: Need to handle potential key errors. Especially for missing description, for which we may be ok with just adding an empty string

File: packages/views.py [hunk 7]
```diff
@@ -3,13 +3,16 @@
...
 
 
 class PackageView(views.APIView):
     renderer_classes = [renderers.JSONRenderer]
 
-    def get(self, request: Request, package_name: str, version: str):
-        package_info = npm.get_versioned_package(package_name, version)
-        serializer = VersionedPackageSerializer(package_info)
+    def get(self, request: Request, package_name: str, range: str | None = None):
+        if range is None:
```
Comment (avivna) [inline packages/views.py:13]: in case of packages like @snyk/snyk-docker-plugin -> range will not be an appropriate version. Need to verify range indeed has expected expression pattern (you can use regular expression), and throw exception in case of unexpected range pattern.

File: tests/test_package.py [hunk 8]
```diff
@@ -3,11 +3,19 @@
 
 def test_get_package():
```
Comment (avivna) [inline tests/test_package.py:4]: This component test is failing, since some of the retrieved packages versions are incorrect.
- need to fix expected versions
- If possible, need to add support in comparing to a version range, and not to a specific version.

File: packages/modules/npm.py [hunk 9]
```diff
@@ -1,21 +1,44 @@
...
+import semver
 
 from packages.models import VersionedPackage
 
 NPM_REGISTRY_URL = "https://registry.npmjs.org"
 
 
-def get_versioned_package(name: str, version: str) -> VersionedPackage:
-    url = f"{NPM_REGISTRY_URL}/{name}/{version}"
+def get_package(name: str, range: str) -> VersionedPackage:
```
Comment (avivna) [inline packages/modules/npm.py:9]: high level comment: This PR does not include any logs, making debugging and error detection difficult. This is particularly true for this function, that includes the main logic of the recursive package data acquisition.

File: tests/test_package.py [hunk 10]
```diff
@@ -3,11 +3,19 @@
```
Comment (avivna) [inline tests/test_package.py:3]: High level comment: There are missing tests for some edge cases like deprecated packages, packages with complex names and circular dependencies.
While the first two (deprecated packages, packages with complex names) can be added as additional component tests, checking for circular dependencies or for cache use (in case cache is added) can be carried in dedicated Unit Tests

File: packages/modules/npm.py (hunk 4 above)
Comment (rajeshpotti-eng) [inline packages/modules/npm.py:12]: Review - Switch to using httpx.AsyncClient for making asynchronous HTTP requests
    # to improve performance.
    #try:
    # async with httpx.AsyncClient() as client:
    #    response = await client.get(f"{NPM_REGISTRY_URL}/{name}")
    #    # Added error handling with response.raise_for_status(). It is good practice
    #    response.raise_for_status() # Raise an exception for HTTP errors
    #   npm_package = response.json()
    # except requests.exceptions.RequestException as e: 
    #    print(f"Error fetching package {name}: {e}")
    #    return {"error": str(e)}

File: packages/models.py (hunk 1 above)
Comment (rajeshpotti-eng) [inline packages/models.py:7]: good improvement for type hinting. It makes the code more readable and maintainable.

File: packages/modules/npm.py (hunk 4 above)
Comment (rajeshpotti-eng) [reply packages/modules/npm.py:12]: Adding version ranges is a great addition. Add error handling when version is not found

File: packages/modules/npm.py [hunk 11]
```diff
@@ -1,21 +1,44 @@
...
+    dependencies = version_record.get("dependencies", {})
+
+    package.dependencies = [
+        get_package(name=dep_name, range=dep_range) for dep_name, dep_range in dependencies.items()
+    ]
 
     return package
+
+
+def request_package(name: str, range: str) -> tuple[VersionedPackage, dict]:
```
Comment (rajeshpotti-eng) [inline packages/modules/npm.py:31]: Change request_package to fetch_package_with_dependencies.

File: packages/serializers.py [hunk 12]
```diff
@@ -1,8 +1,8 @@
...
+from rest_framework_recursive.fields import RecursiveField
 
 
-class VersionedPackageSerializer(serializers.Serializer):
+class PackageSerializer(serializers.Serializer):
     name = serializers.CharField(required=True)
-    description = serializers.CharField(required=True, allow_null=True)
     version = serializers.CharField()
-    dependencies = serializers.DictField(child=serializers.CharField())
+    dependencies = serializers.ListSerializer(child=RecursiveField())
```
Comment (rajeshpotti-eng) [inline packages/serializers.py:8]: Changing to RecursiveField is a good change. It simplifies the serialization of nested dependencies

File: packages/urls.py [hunk 13]
```diff
@@ -4,5 +4,5 @@
 
 urlpatterns = [
     path("<str:package_name>", views.PackageView.as_view()),
-    path("<str:package_name>/<str:version>", views.PackageView.as_view()),
+    path("<str:package_name>/<str:range>", views.PackageView.as_view()),
```
Comment (rajeshpotti-eng) [inline packages/urls.py:7]: Ensure documentation is updated to reflect this change

File: packages/views.py [hunk 14]
```diff
@@ -3,13 +3,16 @@
...
+            range = "*"
```
Comment (rajeshpotti-eng) [inline packages/views.py:14]: Add logging for better debugging and monitoring

File: packages/serializers.py (hunk 12 above)
Comment (rachelcdavies) [reply packages/serializers.py:8]: @rajeshpotti-eng this is a public repo and your comments are visible to all candidates. We normally delete these comments. i am interviewing you later today so we can leave them here until then.

File: packages/modules/npm.py (hunk 4 above)
Comment (rachelcdavies) [reply packages/modules/npm.py:12]: @rajeshpotti-eng this is a public repo and your comments are visible to all candidates. We normally delete these comments. i am interviewing you later today so we can leave them here until then.

File: packages/models.py (hunk 1 above)
Comment (rachelcdavies) [reply packages/models.py:7]: @rajeshpotti-eng this is a public repo and your comments are visible to all candidates. We normally delete these comments. i am interviewing you later today so we can leave them here until then.

File: packages/urls.py (hunk 13 above)
Comment (rachelcdavies) [reply packages/urls.py:7]: @rajeshpotti-eng this is a public repo and your comments are visible to all candidates. We normally delete these comments. i am interviewing you later today so we can leave them here until then.

File: packages/views.py (hunk 14 above)
Comment (rachelcdavies) [reply packages/views.py:14]: @rajeshpotti-eng this is a public repo and your comments are visible to all candidates. We normally delete these comments. i am interviewing you later today so we can leave them here until then.

**Comment by rachelcdavies** [comment]
If you are a candidates who is reviewing this PR as part of an interview, please do **not** add your comments directly into this PR in GitHub because these will be publicly visible to everyone.

File: packages/modules/npm.py (hunk 2 above)
Comment (nicolesolov) [inline packages/modules/npm.py:14]: t

File: packages/views.py [hunk 15]
```diff
@@ -3,13 +3,16 @@
...
+from packages.serializers import PackageSerializer
 
 
 class PackageView(views.APIView):
     renderer_classes = [renderers.JSONRenderer]
 
-    def get(self, request: Request, package_name: str, version: str):
-        package_info = npm.get_versioned_package(package_name, version)
-        serializer = VersionedPackageSerializer(package_info)
+    def get(self, request: Request, package_name: str, range: str | None = None):
```
Comment (keselius) [inline packages/views.py:12]: Request variable is not used - should be removed

File: packages/modules/npm.py [hunk 16]
```diff
@@ -1,21 +1,44 @@
...
+    url = f"{NPM_REGISTRY_URL}/{name}"
+
+    npm_package = requests.get(url).json()
+
+    versions = list(npm_package["versions"].keys())
+    version = semver.min_satisfying(versions, range)
+    version_record = npm_package["versions"][version]
```
Comment (keselius) [inline packages/modules/npm.py:38]: Repeated Code - best to separate out into a helper function

File: packages/views.py (hunk 15 above)
Comment (rachelcdavies) [reply packages/views.py:12]: @keselius if you are interviewing for Snyk please read the instructions. We are not asking you to make comments in a public repo because this is visible to all candidates applying for a job at Snyk. I will be deleting your comments now

File: tests/test_package.py [hunk 17]
```diff
@@ -3,11 +3,19 @@
 
 def test_get_package():
     client = Client()
-    response = client.get("/package/express/2.0.0")
+    response = client.get("/package/minimatch/3.1.2")
     assert response.status_code == 200
     assert response.json() == {
-        "dependencies": {"connect": ">= 1.1.0 < 2.0.0", "mime": ">= 0.0.1", "qs": ">= 0.0.6"},
```
Comment (supanekar15) [inline tests/test_package.py:9]: - For this integration test better to mock the server response insted of contacting real server.
- If talking to npms server is Ok, we should do field level comparison of package names and use semver  instead of comparing actual versions as strings as the upstream dependent versions might change
- Also add unit tests for various scenerios like dependency resolution, package fetching , error conditions etc.

**Review by supanekar15 — COMMENTED** [review]
Initial changes will work but performace, error handling and maintainability  & testability need to be addressed
